*** orig/setup.py	2021-01-03 14:59:39.000000000 -0600
--- new/setup.py	2021-01-03 15:36:19.000000000 -0600
***************
*** 1058,1064 ****
                                                       'termcap'):
                  readline_libs.append('termcap')
              self.add(Extension('readline', ['readline.c'],
!                                library_dirs=['/usr/lib/termcap'],
                                 extra_link_args=readline_extra_link_args,
                                 libraries=readline_libs))
          else:
--- 1058,1064 ----
                                                       'termcap'):
                  readline_libs.append('termcap')
              self.add(Extension('readline', ['readline.c'],
! #                               library_dirs=['/usr/lib/termcap'],
                                 extra_link_args=readline_extra_link_args,
                                 libraries=readline_libs))
          else:
