diff -rc Python-3.12.6.orig/Lib/tkinter/__init__.py Python-3.12.6/Lib/tkinter/__init__.py
*** Python-3.12.6.orig/Lib/tkinter/__init__.py	Fri Sep  6 14:03:47 2024
--- Python-3.12.6/Lib/tkinter/__init__.py	Fri Oct 18 15:10:31 2024
***************
*** 295,301 ****
--- 295,303 ----
              ''.join(' %s=%s' % (k, attrs[k]) for k in keys if k in attrs)
          )
  
+     __class_getitem__ = classmethod(types.GenericAlias)
  
+ 
  _support_default_root = True
  _default_root = None
  
***************
*** 898,907 ****
--- 900,1003 ----
              pass
          self.tk.call('after', 'cancel', id)
  
+     def after_info(self, id=None):
+         """Return information about existing event handlers.
+ 
+         With no argument, return a tuple of the identifiers for all existing
+         event handlers created by the after and after_idle commands for this
+         interpreter.  If id is supplied, it specifies an existing handler; id
+         must have been the return value from some previous call to after or
+         after_idle and it must not have triggered yet or been canceled. If the
+         id doesn't exist, a TclError is raised.  Otherwise, the return value is
+         a tuple containing (script, type) where script is a reference to the
+         function to be called by the event handler and type is either 'idle'
+         or 'timer' to indicate what kind of event handler it is.
+         """
+         return self.tk.splitlist(self.tk.call('after', 'info', id))
+ 
      def bell(self, displayof=0):
          """Ring a display's bell."""
          self.tk.call(('bell',) + self._displayof(displayof))
  
+     def tk_busy_cget(self, option):
+         """Return the value of busy configuration option.
+ 
+         The widget must have been previously made busy by
+         tk_busy_hold().  Option may have any of the values accepted by
+         tk_busy_hold().
+         """
+         return self.tk.call('tk', 'busy', 'cget', self._w, '-'+option)
+     busy_cget = tk_busy_cget
+ 
+     def tk_busy_configure(self, cnf=None, **kw):
+         """Query or modify the busy configuration options.
+ 
+         The widget must have been previously made busy by
+         tk_busy_hold().  Options may have any of the values accepted by
+         tk_busy_hold().
+ 
+         Please note that the option database is referenced by the widget
+         name or class.  For example, if a Frame widget with name "frame"
+         is to be made busy, the busy cursor can be specified for it by
+         either call:
+ 
+             w.option_add('*frame.busyCursor', 'gumby')
+             w.option_add('*Frame.BusyCursor', 'gumby')
+         """
+         if kw:
+             cnf = _cnfmerge((cnf, kw))
+         elif cnf:
+             cnf = _cnfmerge(cnf)
+         if cnf is None:
+             return self._getconfigure(
+                         'tk', 'busy', 'configure', self._w)
+         if isinstance(cnf, str):
+             return self._getconfigure1(
+                         'tk', 'busy', 'configure', self._w, '-'+cnf)
+         self.tk.call('tk', 'busy', 'configure', self._w, *self._options(cnf))
+     busy_config = busy_configure = tk_busy_config = tk_busy_configure
+ 
+     def tk_busy_current(self, pattern=None):
+         """Return a list of widgets that are currently busy.
+ 
+         If a pattern is given, only busy widgets whose path names match
+         a pattern are returned.
+         """
+         return [self._nametowidget(x) for x in
+                 self.tk.splitlist(self.tk.call(
+                    'tk', 'busy', 'current', pattern))]
+     busy_current = tk_busy_current
+ 
+     def tk_busy_forget(self):
+         """Make this widget no longer busy.
+ 
+         User events will again be received by the widget.
+         """
+         self.tk.call('tk', 'busy', 'forget', self._w)
+     busy_forget = tk_busy_forget
+ 
+     def tk_busy_hold(self, **kw):
+         """Make this widget appear busy.
+ 
+         The specified widget and its descendants will be blocked from
+         user interactions.  Normally update() should be called
+         immediately afterward to insure that the hold operation is in
+         effect before the application starts its processing.
+ 
+         The only supported configuration option is:
+ 
+             cursor: the cursor to be displayed when the widget is made
+                     busy.
+         """
+         self.tk.call('tk', 'busy', 'hold', self._w, *self._options(kw))
+     busy = busy_hold = tk_busy = tk_busy_hold
+ 
+     def tk_busy_status(self):
+         """Return True if the widget is busy, False otherwise."""
+         return self.tk.getboolean(self.tk.call(
+                 'tk', 'busy', 'status', self._w))
+     busy_status = tk_busy_status
+ 
      # Clipboard handling:
      def clipboard_get(self, **kw):
          """Retrieve data from the clipboard on window's display.
***************
*** 1633,1638 ****
--- 1729,1737 ----
              except (ValueError, TclError):
                  return s
  
+         if any(isinstance(s, tuple) for s in args):
+             args = [s[0] if isinstance(s, tuple) and len(s) == 1 else s
+                     for s in args]
          nsign, b, f, h, k, s, t, w, x, y, A, E, K, N, W, T, X, Y, D = args
          # Missing: (a, c, d, m, o, v, B, R)
          e = Event()
***************
*** 1668,1674 ****
          try:
              e.type = EventType(T)
          except ValueError:
!             e.type = T
          try:
              e.widget = self._nametowidget(W)
          except KeyError:
--- 1767,1776 ----
          try:
              e.type = EventType(T)
          except ValueError:
!             try:
!                 e.type = EventType(str(T))  # can be int
!             except ValueError:
!                 e.type = T
          try:
              e.widget = self._nametowidget(W)
          except KeyError:
***************
*** 2030,2055 ****
  
      aspect = wm_aspect
  
!     def wm_attributes(self, *args):
!         """This subcommand returns or sets platform specific attributes
  
!         The first form returns a list of the platform specific flags and
!         their values. The second form returns the value for the specific
!         option. The third form sets one or more of the values. The values
!         are as follows:
  
!         On Windows, -disabled gets or sets whether the window is in a
!         disabled state. -toolwindow gets or sets the style of the window
!         to toolwindow (as defined in the MSDN). -topmost gets or sets
!         whether this is a topmost window (displays above all other
!         windows).
! 
!         On Macintosh, XXXXX
! 
!         On Unix, there are currently no special attribute values.
          """
!         args = ('wm', 'attributes', self._w) + args
!         return self.tk.call(args)
  
      attributes = wm_attributes
  
--- 2132,2170 ----
  
      aspect = wm_aspect
  
!     def wm_attributes(self, *args, return_python_dict=False, **kwargs):
!         """Return or sets platform specific attributes.
  
!         When called with a single argument return_python_dict=True,
!         return a dict of the platform specific attributes and their values.
!         When called without arguments or with a single argument
!         return_python_dict=False, return a tuple containing intermixed
!         attribute names with the minus prefix and their values.
  
!         When called with a single string value, return the value for the
!         specific option.  When called with keyword arguments, set the
!         corresponding attributes.
          """
!         if not kwargs:
!             if not args:
!                 res = self.tk.call('wm', 'attributes', self._w)
!                 if return_python_dict:
!                     return _splitdict(self.tk, res)
!                 else:
!                     return self.tk.splitlist(res)
!             if len(args) == 1 and args[0] is not None:
!                 option = args[0]
!                 if option[0] == '-':
!                     # TODO: deprecate
!                     option = option[1:]
!                 return self.tk.call('wm', 'attributes', self._w, '-' + option)
!             # TODO: deprecate
!             return self.tk.call('wm', 'attributes', self._w, *args)
!         elif args:
!             raise TypeError('wm_attribute() options have been specified as '
!                             'positional and keyword arguments')
!         else:
!             self.tk.call('wm', 'attributes', self._w, *self._options(kwargs))
  
      attributes = wm_attributes
  
***************
*** 3669,3693 ****
          return self.tk.getboolean(self.tk.call(
              self._w, 'compare', index1, op, index2))
  
!     def count(self, index1, index2, *args): # new in Tk 8.5
          """Counts the number of relevant things between the two indices.
!         If index1 is after index2, the result will be a negative number
          (and this holds for each of the possible options).
  
!         The actual items which are counted depends on the options given by
!         args. The result is a list of integers, one for the result of each
!         counting option given. Valid counting options are "chars",
!         "displaychars", "displayindices", "displaylines", "indices",
!         "lines", "xpixels" and "ypixels". There is an additional possible
!         option "update", which if given then all subsequent options ensure
!         that any possible out of date information is recalculated."""
!         args = ['-%s' % arg for arg in args]
!         args += [index1, index2]
!         res = self.tk.call(self._w, 'count', *args) or None
!         if res is not None and len(args) <= 3:
!             return (res, )
!         else:
!             return res
  
      def debug(self, boolean=None):
          """Turn on the internal consistency checks of the B-Tree inside the text
--- 3784,3818 ----
          return self.tk.getboolean(self.tk.call(
              self._w, 'compare', index1, op, index2))
  
!     def count(self, index1, index2, *options, return_ints=False): # new in Tk 8.5
          """Counts the number of relevant things between the two indices.
! 
!         If INDEX1 is after INDEX2, the result will be a negative number
          (and this holds for each of the possible options).
  
!         The actual items which are counted depends on the options given.
!         The result is a tuple of integers, one for the result of each
!         counting option given, if more than one option is specified or
!         return_ints is false (default), otherwise it is an integer.
!         Valid counting options are "chars", "displaychars",
!         "displayindices", "displaylines", "indices", "lines", "xpixels"
!         and "ypixels". The default value, if no option is specified, is
!         "indices". There is an additional possible option "update",
!         which if given then all subsequent options ensure that any
!         possible out of date information is recalculated.
!         """
!         options = ['-%s' % arg for arg in options]
!         res = self.tk.call(self._w, 'count', *options, index1, index2)
!         if not isinstance(res, int):
!             res = self._getints(res)
!             if len(res) == 1:
!                 res, = res
!         if not return_ints:
!             if not res:
!                 res = None
!             elif len(options) <= 1:
!                 res = (res,)
!         return res
  
      def debug(self, boolean=None):
          """Turn on the internal consistency checks of the B-Tree inside the text
***************
*** 4172,4205 ****
  
      def __getitem__(self, key):
          return self.tk.call(self.name, 'cget', '-' + key)
-     # XXX copy -from, -to, ...?
  
!     def copy(self):
!         """Return a new PhotoImage with the same image as this widget."""
          destImage = PhotoImage(master=self.tk)
!         self.tk.call(destImage, 'copy', self.name)
          return destImage
  
!     def zoom(self, x, y=''):
          """Return a new PhotoImage with the same image as this widget
!         but zoom it with a factor of x in the X direction and y in the Y
!         direction.  If y is not given, the default value is the same as x.
          """
-         destImage = PhotoImage(master=self.tk)
          if y=='': y=x
!         self.tk.call(destImage, 'copy', self.name, '-zoom',x,y)
!         return destImage
  
!     def subsample(self, x, y=''):
          """Return a new PhotoImage based on the same image as this widget
!         but use only every Xth or Yth pixel.  If y is not given, the
!         default value is the same as x.
          """
-         destImage = PhotoImage(master=self.tk)
          if y=='': y=x
!         self.tk.call(destImage, 'copy', self.name, '-subsample',x,y)
!         return destImage
  
      def get(self, x, y):
          """Return the color (red, green, blue) of the pixel at X,Y."""
          return self.tk.call(self.name, 'get', x, y)
--- 4297,4409 ----
  
      def __getitem__(self, key):
          return self.tk.call(self.name, 'cget', '-' + key)
  
!     def copy(self, *, from_coords=None, zoom=None, subsample=None):
!         """Return a new PhotoImage with the same image as this widget.
! 
!         The FROM_COORDS option specifies a rectangular sub-region of the
!         source image to be copied. It must be a tuple or a list of 1 to 4
!         integers (x1, y1, x2, y2).  (x1, y1) and (x2, y2) specify diagonally
!         opposite corners of the rectangle.  If x2 and y2 are not specified,
!         the default value is the bottom-right corner of the source image.
!         The pixels copied will include the left and top edges of the
!         specified rectangle but not the bottom or right edges.  If the
!         FROM_COORDS option is not given, the default is the whole source
!         image.
! 
!         If SUBSAMPLE or ZOOM are specified, the image is transformed as in
!         the subsample() or zoom() methods.  The value must be a single
!         integer or a pair of integers.
!         """
          destImage = PhotoImage(master=self.tk)
!         destImage.copy_replace(self, from_coords=from_coords,
!                                zoom=zoom, subsample=subsample)
          return destImage
  
!     def zoom(self, x, y='', *, from_coords=None):
          """Return a new PhotoImage with the same image as this widget
!         but zoom it with a factor of X in the X direction and Y in the Y
!         direction.  If Y is not given, the default value is the same as X.
! 
!         The FROM_COORDS option specifies a rectangular sub-region of the
!         source image to be copied, as in the copy() method.
          """
          if y=='': y=x
!         return self.copy(zoom=(x, y), from_coords=from_coords)
  
!     def subsample(self, x, y='', *, from_coords=None):
          """Return a new PhotoImage based on the same image as this widget
!         but use only every Xth or Yth pixel.  If Y is not given, the
!         default value is the same as X.
! 
!         The FROM_COORDS option specifies a rectangular sub-region of the
!         source image to be copied, as in the copy() method.
          """
          if y=='': y=x
!         return self.copy(subsample=(x, y), from_coords=from_coords)
  
+     def copy_replace(self, sourceImage, *, from_coords=None, to=None, shrink=False,
+                      zoom=None, subsample=None, compositingrule=None):
+         """Copy a region from the source image (which must be a PhotoImage) to
+         this image, possibly with pixel zooming and/or subsampling.  If no
+         options are specified, this command copies the whole of the source
+         image into this image, starting at coordinates (0, 0).
+ 
+         The FROM_COORDS option specifies a rectangular sub-region of the
+         source image to be copied. It must be a tuple or a list of 1 to 4
+         integers (x1, y1, x2, y2).  (x1, y1) and (x2, y2) specify diagonally
+         opposite corners of the rectangle.  If x2 and y2 are not specified,
+         the default value is the bottom-right corner of the source image.
+         The pixels copied will include the left and top edges of the
+         specified rectangle but not the bottom or right edges.  If the
+         FROM_COORDS option is not given, the default is the whole source
+         image.
+ 
+         The TO option specifies a rectangular sub-region of the destination
+         image to be affected.  It must be a tuple or a list of 1 to 4
+         integers (x1, y1, x2, y2).  (x1, y1) and (x2, y2) specify diagonally
+         opposite corners of the rectangle.  If x2 and y2 are not specified,
+         the default value is (x1,y1) plus the size of the source region
+         (after subsampling and zooming, if specified).  If x2 and y2 are
+         specified, the source region will be replicated if necessary to fill
+         the destination region in a tiled fashion.
+ 
+         If SHRINK is true, the size of the destination image should be
+         reduced, if necessary, so that the region being copied into is at
+         the bottom-right corner of the image.
+ 
+         If SUBSAMPLE or ZOOM are specified, the image is transformed as in
+         the subsample() or zoom() methods.  The value must be a single
+         integer or a pair of integers.
+ 
+         The COMPOSITINGRULE option specifies how transparent pixels in the
+         source image are combined with the destination image.  When a
+         compositing rule of 'overlay' is set, the old contents of the
+         destination image are visible, as if the source image were printed
+         on a piece of transparent film and placed over the top of the
+         destination.  When a compositing rule of 'set' is set, the old
+         contents of the destination image are discarded and the source image
+         is used as-is.  The default compositing rule is 'overlay'.
+         """
+         options = []
+         if from_coords is not None:
+             options.extend(('-from', *from_coords))
+         if to is not None:
+             options.extend(('-to', *to))
+         if shrink:
+             options.append('-shrink')
+         if zoom is not None:
+             if not isinstance(zoom, (tuple, list)):
+                 zoom = (zoom,)
+             options.extend(('-zoom', *zoom))
+         if subsample is not None:
+             if not isinstance(subsample, (tuple, list)):
+                 subsample = (subsample,)
+             options.extend(('-subsample', *subsample))
+         if compositingrule:
+             options.extend(('-compositingrule', compositingrule))
+         self.tk.call(self.name, 'copy', sourceImage, *options)
+ 
      def get(self, x, y):
          """Return the color (red, green, blue) of the pixel at X,Y."""
          return self.tk.call(self.name, 'get', x, y)
***************
*** 4213,4229 ****
                  to = to[1:]
              args = args + ('-to',) + tuple(to)
          self.tk.call(args)
-     # XXX read
  
!     def write(self, filename, format=None, from_coords=None):
!         """Write image to file FILENAME in FORMAT starting from
!         position FROM_COORDS."""
!         args = (self.name, 'write', filename)
!         if format:
!             args = args + ('-format', format)
!         if from_coords:
!             args = args + ('-from',) + tuple(from_coords)
!         self.tk.call(args)
  
      def transparency_get(self, x, y):
          """Return True if the pixel at x,y is transparent."""
--- 4417,4533 ----
                  to = to[1:]
              args = args + ('-to',) + tuple(to)
          self.tk.call(args)
  
!     def read(self, filename, format=None, *, from_coords=None, to=None, shrink=False):
!         """Reads image data from the file named FILENAME into the image.
! 
!         The FORMAT option specifies the format of the image data in the
!         file.
! 
!         The FROM_COORDS option specifies a rectangular sub-region of the image
!         file data to be copied to the destination image.  It must be a tuple
!         or a list of 1 to 4 integers (x1, y1, x2, y2).  (x1, y1) and
!         (x2, y2) specify diagonally opposite corners of the rectangle.  If
!         x2 and y2 are not specified, the default value is the bottom-right
!         corner of the source image.  The default, if this option is not
!         specified, is the whole of the image in the image file.
! 
!         The TO option specifies the coordinates of the top-left corner of
!         the region of the image into which data from filename are to be
!         read.  The default is (0, 0).
! 
!         If SHRINK is true, the size of the destination image will be
!         reduced, if necessary, so that the region into which the image file
!         data are read is at the bottom-right corner of the image.
!         """
!         options = ()
!         if format is not None:
!             options += ('-format', format)
!         if from_coords is not None:
!             options += ('-from', *from_coords)
!         if shrink:
!             options += ('-shrink',)
!         if to is not None:
!             options += ('-to', *to)
!         self.tk.call(self.name, 'read', filename, *options)
! 
!     def write(self, filename, format=None, from_coords=None, *,
!               background=None, grayscale=False):
!         """Writes image data from the image to a file named FILENAME.
! 
!         The FORMAT option specifies the name of the image file format
!         handler to be used to write the data to the file.  If this option
!         is not given, the format is guessed from the file extension.
! 
!         The FROM_COORDS option specifies a rectangular region of the image
!         to be written to the image file.  It must be a tuple or a list of 1
!         to 4 integers (x1, y1, x2, y2).  If only x1 and y1 are specified,
!         the region extends from (x1,y1) to the bottom-right corner of the
!         image.  If all four coordinates are given, they specify diagonally
!         opposite corners of the rectangular region.  The default, if this
!         option is not given, is the whole image.
! 
!         If BACKGROUND is specified, the data will not contain any
!         transparency information.  In all transparent pixels the color will
!         be replaced by the specified color.
! 
!         If GRAYSCALE is true, the data will not contain color information.
!         All pixel data will be transformed into grayscale.
!         """
!         options = ()
!         if format is not None:
!             options += ('-format', format)
!         if from_coords is not None:
!             options += ('-from', *from_coords)
!         if grayscale:
!             options += ('-grayscale',)
!         if background is not None:
!             options += ('-background', background)
!         self.tk.call(self.name, 'write', filename, *options)
! 
!     def data(self, format=None, *, from_coords=None,
!              background=None, grayscale=False):
!         """Returns image data.
! 
!         The FORMAT option specifies the name of the image file format
!         handler to be used.  If this option is not given, this method uses
!         a format that consists of a tuple (one element per row) of strings
!         containing space-separated (one element per pixel/column) colors
!         in “#RRGGBB” format (where RR is a pair of hexadecimal digits for
!         the red channel, GG for green, and BB for blue).
! 
!         The FROM_COORDS option specifies a rectangular region of the image
!         to be returned.  It must be a tuple or a list of 1 to 4 integers
!         (x1, y1, x2, y2).  If only x1 and y1 are specified, the region
!         extends from (x1,y1) to the bottom-right corner of the image.  If
!         all four coordinates are given, they specify diagonally opposite
!         corners of the rectangular region, including (x1, y1) and excluding
!         (x2, y2).  The default, if this option is not given, is the whole
!         image.
! 
!         If BACKGROUND is specified, the data will not contain any
!         transparency information.  In all transparent pixels the color will
!         be replaced by the specified color.
! 
!         If GRAYSCALE is true, the data will not contain color information.
!         All pixel data will be transformed into grayscale.
!         """
!         options = ()
!         if format is not None:
!             options += ('-format', format)
!         if from_coords is not None:
!             options += ('-from', *from_coords)
!         if grayscale:
!             options += ('-grayscale',)
!         if background is not None:
!             options += ('-background', background)
!         data = self.tk.call(self.name, 'data', *options)
!         if isinstance(data, str):  # For wantobjects = 0.
!             if format is None:
!                 data = self.tk.splitlist(data)
!             else:
!                 data = bytes(data, 'latin1')
!         return data
  
      def transparency_get(self, x, y):
          """Return True if the pixel at x,y is transparent."""
diff -rc Python-3.12.6.orig/Modules/_tkinter.c Python-3.12.6/Modules/_tkinter.c
*** Python-3.12.6.orig/Modules/_tkinter.c	Fri Sep  6 14:03:47 2024
--- Python-3.12.6/Modules/_tkinter.c	Fri Oct 18 15:09:41 2024
***************
*** 21,41 ****
  
  */
  
- #define PY_SSIZE_T_CLEAN
  #ifndef Py_BUILD_CORE_BUILTIN
  #  define Py_BUILD_CORE_MODULE 1
  #endif
  
  #include "Python.h"
- #include <ctype.h>
  #ifdef MS_WINDOWS
  #  include "pycore_fileutils.h"   // _Py_stat()
  #endif
  
! #include "pycore_long.h"
  
  #ifdef MS_WINDOWS
! #include <windows.h>
  #endif
  
  #define CHECK_SIZE(size, elemsize) \
--- 21,39 ----
  
  */
  
  #ifndef Py_BUILD_CORE_BUILTIN
  #  define Py_BUILD_CORE_MODULE 1
  #endif
  
  #include "Python.h"
  #ifdef MS_WINDOWS
  #  include "pycore_fileutils.h"   // _Py_stat()
  #endif
  
! #include "pycore_long.h"          // _PyLong_IsNegative()
  
  #ifdef MS_WINDOWS
! #  include <windows.h>
  #endif
  
  #define CHECK_SIZE(size, elemsize) \
***************
*** 47,57 ****
  #define TCL_THREADS
  
  #ifdef TK_FRAMEWORK
! #include <Tcl/tcl.h>
! #include <Tk/tk.h>
  #else
! #include <tcl.h>
! #include <tk.h>
  #endif
  
  #include "tkinter.h"
--- 45,55 ----
  #define TCL_THREADS
  
  #ifdef TK_FRAMEWORK
! #  include <Tcl/tcl.h>
! #  include <Tk/tk.h>
  #else
! #  include <tcl.h>
! #  include <tk.h>
  #endif
  
  #include "tkinter.h"
***************
*** 142,152 ****
      static int already_checked = 0;
  
      if (already_checked == 0) {
-         PyObject *prefix;
          struct stat stat_buf;
          int stat_return_value;
  
!         prefix = PyUnicode_FromWideChar(Py_GetPrefix(), -1);
          if (prefix == NULL) {
              return NULL;
          }
--- 140,149 ----
      static int already_checked = 0;
  
      if (already_checked == 0) {
          struct stat stat_buf;
          int stat_return_value;
  
!         PyObject *prefix = PySys_GetObject("prefix");  // borrowed reference
          if (prefix == NULL) {
              return NULL;
          }
***************
*** 326,334 ****
      const Tcl_ObjType *WideIntType;
      const Tcl_ObjType *BignumType;
      const Tcl_ObjType *ListType;
-     const Tcl_ObjType *ProcBodyType;
      const Tcl_ObjType *StringType;
      const Tcl_ObjType *UTF32StringType;
  } TkappObject;
  
  #define Tkapp_Interp(v) (((TkappObject *) (v))->interp)
--- 323,331 ----
      const Tcl_ObjType *WideIntType;
      const Tcl_ObjType *BignumType;
      const Tcl_ObjType *ListType;
      const Tcl_ObjType *StringType;
      const Tcl_ObjType *UTF32StringType;
+     const Tcl_ObjType *PixelType;
  } TkappObject;
  
  #define Tkapp_Interp(v) (((TkappObject *) (v))->interp)
***************
*** 639,647 ****
      v->WideIntType = Tcl_GetObjType("wideInt");
      v->BignumType = Tcl_GetObjType("bignum");
      v->ListType = Tcl_GetObjType("list");
-     v->ProcBodyType = Tcl_GetObjType("procbody");
      v->StringType = Tcl_GetObjType("string");
      v->UTF32StringType = Tcl_GetObjType("utf32string");
  
      /* Delete the 'exit' command, which can screw things up */
      Tcl_DeleteCommand(v->interp, "exit");
--- 636,644 ----
      v->WideIntType = Tcl_GetObjType("wideInt");
      v->BignumType = Tcl_GetObjType("bignum");
      v->ListType = Tcl_GetObjType("list");
      v->StringType = Tcl_GetObjType("string");
      v->UTF32StringType = Tcl_GetObjType("utf32string");
+     v->PixelType = Tcl_GetObjType("pixel");
  
      /* Delete the 'exit' command, which can screw things up */
      Tcl_DeleteCommand(v->interp, "exit");
***************
*** 783,790 ****
  }
  
  static void
! PyTclObject_dealloc(PyTclObject *self)
  {
      PyObject *tp = (PyObject *) Py_TYPE(self);
      Tcl_DecrRefCount(self->value);
      Py_XDECREF(self->string);
--- 780,788 ----
  }
  
  static void
! PyTclObject_dealloc(PyObject *_self)
  {
+     PyTclObject *self = (PyTclObject *)_self;
      PyObject *tp = (PyObject *) Py_TYPE(self);
      Tcl_DecrRefCount(self->value);
      Py_XDECREF(self->string);
***************
*** 797,804 ****
  "the string representation of this object, either as str or bytes");
  
  static PyObject *
! PyTclObject_string(PyTclObject *self, void *ignored)
  {
      if (!self->string) {
          self->string = unicodeFromTclObj(NULL, self->value);
          if (!self->string)
--- 795,803 ----
  "the string representation of this object, either as str or bytes");
  
  static PyObject *
! PyTclObject_string(PyObject *_self, void *ignored)
  {
+     PyTclObject *self = (PyTclObject *)_self;
      if (!self->string) {
          self->string = unicodeFromTclObj(NULL, self->value);
          if (!self->string)
***************
*** 808,815 ****
  }
  
  static PyObject *
! PyTclObject_str(PyTclObject *self)
  {
      if (self->string) {
          return Py_NewRef(self->string);
      }
--- 807,815 ----
  }
  
  static PyObject *
! PyTclObject_str(PyObject *_self)
  {
+     PyTclObject *self = (PyTclObject *)_self;
      if (self->string) {
          return Py_NewRef(self->string);
      }
***************
*** 818,826 ****
  }
  
  static PyObject *
! PyTclObject_repr(PyTclObject *self)
  {
!     PyObject *repr, *str = PyTclObject_str(self);
      if (str == NULL)
          return NULL;
      repr = PyUnicode_FromFormat("<%s object: %R>",
--- 818,827 ----
  }
  
  static PyObject *
! PyTclObject_repr(PyObject *_self)
  {
!     PyTclObject *self = (PyTclObject *)_self;
!     PyObject *repr, *str = PyTclObject_str(_self);
      if (str == NULL)
          return NULL;
      repr = PyUnicode_FromFormat("<%s object: %R>",
***************
*** 857,879 ****
  PyDoc_STRVAR(get_typename__doc__, "name of the Tcl type");
  
  static PyObject*
! get_typename(PyTclObject* obj, void* ignored)
  {
      return unicodeFromTclString(obj->value->typePtr->name);
  }
  
  
  static PyGetSetDef PyTclObject_getsetlist[] = {
!     {"typename", (getter)get_typename, NULL, get_typename__doc__},
!     {"string", (getter)PyTclObject_string, NULL,
       PyTclObject_string__doc__},
      {0},
  };
  
  static PyType_Slot PyTclObject_Type_slots[] = {
!     {Py_tp_dealloc, (destructor)PyTclObject_dealloc},
!     {Py_tp_repr, (reprfunc)PyTclObject_repr},
!     {Py_tp_str, (reprfunc)PyTclObject_str},
      {Py_tp_getattro, PyObject_GenericGetAttr},
      {Py_tp_richcompare, PyTclObject_richcompare},
      {Py_tp_getset, PyTclObject_getsetlist},
--- 858,881 ----
  PyDoc_STRVAR(get_typename__doc__, "name of the Tcl type");
  
  static PyObject*
! get_typename(PyObject *self, void* ignored)
  {
+     PyTclObject *obj = (PyTclObject *)self;
      return unicodeFromTclString(obj->value->typePtr->name);
  }
  
  
  static PyGetSetDef PyTclObject_getsetlist[] = {
!     {"typename", get_typename, NULL, get_typename__doc__},
!     {"string", PyTclObject_string, NULL,
       PyTclObject_string__doc__},
      {0},
  };
  
  static PyType_Slot PyTclObject_Type_slots[] = {
!     {Py_tp_dealloc, PyTclObject_dealloc},
!     {Py_tp_repr, PyTclObject_repr},
!     {Py_tp_str, PyTclObject_str},
      {Py_tp_getattro, PyObject_GenericGetAttr},
      {Py_tp_richcompare, PyTclObject_richcompare},
      {Py_tp_getset, PyTclObject_getsetlist},
***************
*** 958,973 ****
      if (PyLong_CheckExact(value)) {
          int overflow;
          long longValue;
- #ifdef TCL_WIDE_INT_TYPE
          Tcl_WideInt wideValue;
- #endif
          longValue = PyLong_AsLongAndOverflow(value, &overflow);
          if (!overflow) {
              return Tcl_NewLongObj(longValue);
          }
          /* If there is an overflow in the long conversion,
             fall through to wideInt handling. */
- #ifdef TCL_WIDE_INT_TYPE
          if (_PyLong_AsByteArray((PyLongObject *)value,
                                  (unsigned char *)(void *)&wideValue,
                                  sizeof(wideValue),
--- 960,972 ----
***************
*** 976,982 ****
              return Tcl_NewWideIntObj(wideValue);
          }
          PyErr_Clear();
- #endif
          /* If there is an overflow in the wideInt conversion,
             fall through to bignum handling. */
          return asBignumObj(value);
--- 975,980 ----
***************
*** 1013,1021 ****
      }
  
      if (PyUnicode_Check(value)) {
-         if (PyUnicode_READY(value) == -1)
-             return NULL;
- 
          Py_ssize_t size = PyUnicode_GET_LENGTH(value);
          if (size == 0) {
              return Tcl_NewStringObj("", 0);
--- 1011,1016 ----
***************
*** 1242,1248 ****
      }
  
      if (value->typePtr == tkapp->StringType ||
!         value->typePtr == tkapp->UTF32StringType)
      {
          return unicodeFromTclObj(tkapp, value);
      }
--- 1237,1244 ----
      }
  
      if (value->typePtr == tkapp->StringType ||
!         value->typePtr == tkapp->UTF32StringType ||
! 	value->typePtr == tkapp->PixelType)
      {
          return unicodeFromTclObj(tkapp, value);
      }
***************
*** 1397,1404 ****
     hold the Python lock. */
  
  static int
! Tkapp_CallProc(Tkapp_CallEvent *e, int flags)
  {
      Tcl_Obj *objStore[ARGSZ];
      Tcl_Obj **objv;
      Tcl_Size objc;
--- 1393,1401 ----
     hold the Python lock. */
  
  static int
! Tkapp_CallProc(Tcl_Event *evPtr, int flags)
  {
+     Tkapp_CallEvent *e = (Tkapp_CallEvent *)evPtr;
      Tcl_Obj *objStore[ARGSZ];
      Tcl_Obj **objv;
      Tcl_Size objc;
***************
*** 1473,1479 ****
             marshal the parameters to the interpreter thread. */
          Tkapp_CallEvent *ev;
          Tcl_Condition cond = NULL;
!         PyObject *exc;
          if (!WaitForMainloop(self))
              return NULL;
          ev = (Tkapp_CallEvent*)attemptckalloc(sizeof(Tkapp_CallEvent));
--- 1470,1476 ----
             marshal the parameters to the interpreter thread. */
          Tkapp_CallEvent *ev;
          Tcl_Condition cond = NULL;
!         PyObject *exc = NULL;  // init to make static analyzers happy
          if (!WaitForMainloop(self))
              return NULL;
          ev = (Tkapp_CallEvent*)attemptckalloc(sizeof(Tkapp_CallEvent));
***************
*** 1481,1487 ****
              PyErr_NoMemory();
              return NULL;
          }
!         ev->ev.proc = (Tcl_EventProc*)Tkapp_CallProc;
          ev->self = self;
          ev->args = args;
          ev->res = &res;
--- 1478,1484 ----
              PyErr_NoMemory();
              return NULL;
          }
!         ev->ev.proc = Tkapp_CallProc;
          ev->self = self;
          ev->args = args;
          ev->res = &res;
***************
*** 1728,1735 ****
  }
  
  static int
! var_proc(VarEvent* ev, int flags)
  {
      ENTER_PYTHON
      var_perform(ev);
      Tcl_MutexLock(&var_mutex);
--- 1725,1733 ----
  }
  
  static int
! var_proc(Tcl_Event *evPtr, int flags)
  {
+     VarEvent *ev = (VarEvent *)evPtr;
      ENTER_PYTHON
      var_perform(ev);
      Tcl_MutexLock(&var_mutex);
***************
*** 1746,1752 ****
      TkappObject *self = (TkappObject*)selfptr;
      if (self->threaded && self->thread_id != Tcl_GetCurrentThread()) {
          VarEvent *ev;
!         PyObject *res, *exc;
          Tcl_Condition cond = NULL;
  
          /* The current thread is not the interpreter thread.  Marshal
--- 1744,1751 ----
      TkappObject *self = (TkappObject*)selfptr;
      if (self->threaded && self->thread_id != Tcl_GetCurrentThread()) {
          VarEvent *ev;
!         // init 'res' and 'exc' to make static analyzers happy
!         PyObject *res = NULL, *exc = NULL;
          Tcl_Condition cond = NULL;
  
          /* The current thread is not the interpreter thread.  Marshal
***************
*** 1767,1773 ****
          ev->res = &res;
          ev->exc = &exc;
          ev->cond = &cond;
!         ev->ev.proc = (Tcl_EventProc*)var_proc;
          Tkapp_ThreadSend(self, (Tcl_Event*)ev, &cond, &var_mutex);
          Tcl_ConditionFinalize(&cond);
          if (!res) {
--- 1766,1772 ----
          ev->res = &res;
          ev->exc = &exc;
          ev->cond = &cond;
!         ev->ev.proc = var_proc;
          Tkapp_ThreadSend(self, (Tcl_Event*)ev, &cond, &var_mutex);
          Tcl_ConditionFinalize(&cond);
          if (!res) {
***************
*** 2306,2312 ****
  
  /* Client data struct */
  typedef struct {
!     PyObject *self;
      PyObject *func;
  } PythonCmd_ClientData;
  
--- 2305,2311 ----
  
  /* Client data struct */
  typedef struct {
!     TkappObject *self;
      PyObject *func;
  } PythonCmd_ClientData;
  
***************
*** 2330,2335 ****
--- 2329,2335 ----
      PyObject *args, *res;
      int i;
      Tcl_Obj *obj_res;
+     int objargs = data->self->wantobjects >= 2;
  
      ENTER_PYTHON
  
***************
*** 2338,2344 ****
          return PythonCmd_Error(interp);
  
      for (i = 0; i < (objc - 1); i++) {
!         PyObject *s = unicodeFromTclObj((TkappObject *)data->self, objv[i + 1]);
          if (!s) {
              Py_DECREF(args);
              return PythonCmd_Error(interp);
--- 2338,2345 ----
          return PythonCmd_Error(interp);
  
      for (i = 0; i < (objc - 1); i++) {
!         PyObject *s = objargs ? FromObj(data->self, objv[i + 1])
!                               : unicodeFromTclObj(data->self, objv[i + 1]);
          if (!s) {
              Py_DECREF(args);
              return PythonCmd_Error(interp);
***************
*** 2394,2401 ****
  } CommandEvent;
  
  static int
! Tkapp_CommandProc(CommandEvent *ev, int flags)
  {
      if (ev->create)
          *ev->status = Tcl_CreateObjCommand(
              ev->interp, ev->name, PythonCmd,
--- 2395,2403 ----
  } CommandEvent;
  
  static int
! Tkapp_CommandProc(Tcl_Event *evPtr, int flags)
  {
+     CommandEvent *ev = (CommandEvent *)evPtr;
      if (ev->create)
          *ev->status = Tcl_CreateObjCommand(
              ev->interp, ev->name, PythonCmd,
***************
*** 2440,2448 ****
      data = PyMem_NEW(PythonCmd_ClientData, 1);
      if (!data)
          return PyErr_NoMemory();
!     data->self = Py_NewRef(self);
      data->func = Py_NewRef(func);
      if (self->threaded && self->thread_id != Tcl_GetCurrentThread()) {
          Tcl_Condition cond = NULL;
          CommandEvent *ev = (CommandEvent*)attemptckalloc(sizeof(CommandEvent));
          if (ev == NULL) {
--- 2442,2453 ----
      data = PyMem_NEW(PythonCmd_ClientData, 1);
      if (!data)
          return PyErr_NoMemory();
!     Py_INCREF(self);
!     data->self = self;
      data->func = Py_NewRef(func);
      if (self->threaded && self->thread_id != Tcl_GetCurrentThread()) {
+         err = 0;  // init to make static analyzers happy
+ 
          Tcl_Condition cond = NULL;
          CommandEvent *ev = (CommandEvent*)attemptckalloc(sizeof(CommandEvent));
          if (ev == NULL) {
***************
*** 2450,2456 ****
              PyMem_Free(data);
              return NULL;
          }
!         ev->ev.proc = (Tcl_EventProc*)Tkapp_CommandProc;
          ev->interp = self->interp;
          ev->create = 1;
          ev->name = name;
--- 2455,2461 ----
              PyMem_Free(data);
              return NULL;
          }
!         ev->ev.proc = Tkapp_CommandProc;
          ev->interp = self->interp;
          ev->create = 1;
          ev->name = name;
***************
*** 2498,2503 ****
--- 2503,2510 ----
      TRACE(self, ("((sss))", "rename", name, ""));
  
      if (self->threaded && self->thread_id != Tcl_GetCurrentThread()) {
+         err = 0;  // init to make static analyzers happy
+ 
          Tcl_Condition cond = NULL;
          CommandEvent *ev;
          ev = (CommandEvent*)attemptckalloc(sizeof(CommandEvent));
***************
*** 2505,2511 ****
              PyErr_NoMemory();
              return NULL;
          }
!         ev->ev.proc = (Tcl_EventProc*)Tkapp_CommandProc;
          ev->interp = self->interp;
          ev->create = 0;
          ev->name = name;
--- 2512,2518 ----
              PyErr_NoMemory();
              return NULL;
          }
!         ev->ev.proc = Tkapp_CommandProc;
          ev->interp = self->interp;
          ev->create = 0;
          ev->name = name;
***************
*** 2954,2963 ****
  {
  
      int wantobjects = -1;
!     if (!PyArg_ParseTuple(args, "|p:wantobjects", &wantobjects))
          return NULL;
      if (wantobjects == -1)
!         return PyBool_FromLong(((TkappObject*)self)->wantobjects);
      ((TkappObject*)self)->wantobjects = wantobjects;
  
      Py_RETURN_NONE;
--- 2961,2970 ----
  {
  
      int wantobjects = -1;
!     if (!PyArg_ParseTuple(args, "|i:wantobjects", &wantobjects))
          return NULL;
      if (wantobjects == -1)
!         return PyLong_FromLong(((TkappObject*)self)->wantobjects);
      ((TkappObject*)self)->wantobjects = wantobjects;
  
      Py_RETURN_NONE;
***************
*** 3143,3149 ****
      baseName: str = ""
      className: str = "Tk"
      interactive: bool = False
!     wantobjects: bool = False
      wantTk: bool = True
          if false, then Tk_Init() doesn't get called
      sync: bool = False
--- 3150,3156 ----
      baseName: str = ""
      className: str = "Tk"
      interactive: bool = False
!     wantobjects: int = 0
      wantTk: bool = True
          if false, then Tk_Init() doesn't get called
      sync: bool = False
***************
*** 3159,3165 ****
                       const char *baseName, const char *className,
                       int interactive, int wantobjects, int wantTk, int sync,
                       const char *use)
! /*[clinic end generated code: output=e3315607648e6bb4 input=09afef9adea70a19]*/
  {
      /* XXX baseName is not used anymore;
       * try getting rid of it. */
--- 3166,3172 ----
                       const char *baseName, const char *className,
                       int interactive, int wantobjects, int wantTk, int sync,
                       const char *use)
! /*[clinic end generated code: output=e3315607648e6bb4 input=7e382ba431bed537]*/
  {
      /* XXX baseName is not used anymore;
       * try getting rid of it. */
***************
*** 3384,3406 ****
  #endif
  }
  
  
  static struct PyModuleDef _tkintermodule = {
      PyModuleDef_HEAD_INIT,
!     "_tkinter",
!     NULL,
!     -1,
!     moduleMethods,
!     NULL,
!     NULL,
!     NULL,
!     NULL
  };
  
  PyMODINIT_FUNC
  PyInit__tkinter(void)
  {
!   PyObject *m, *uexe, *cexe, *o;
  
      tcl_lock = PyThread_allocate_lock();
      if (tcl_lock == NULL)
--- 3391,3436 ----
  #endif
  }
  
+ static int
+ module_clear(PyObject *Py_UNUSED(mod))
+ {
+     Py_CLEAR(Tkinter_TclError);
+     Py_CLEAR(Tkapp_Type);
+     Py_CLEAR(Tktt_Type);
+     Py_CLEAR(PyTclObject_Type);
+     return 0;
+ }
  
+ static int
+ module_traverse(PyObject *Py_UNUSED(module), visitproc visit, void *arg)
+ {
+     Py_VISIT(Tkinter_TclError);
+     Py_VISIT(Tkapp_Type);
+     Py_VISIT(Tktt_Type);
+     Py_VISIT(PyTclObject_Type);
+     return 0;
+ }
+ 
+ static void
+ module_free(void *mod)
+ {
+     (void)module_clear((PyObject *)mod);
+ }
+ 
  static struct PyModuleDef _tkintermodule = {
      PyModuleDef_HEAD_INIT,
!     .m_name = "_tkinter",
!     .m_size = -1,
!     .m_methods = moduleMethods,
!     .m_traverse = module_traverse,
!     .m_clear = module_clear,
!     .m_free = module_free
  };
  
  PyMODINIT_FUNC
  PyInit__tkinter(void)
  {
!     PyObject *m, *uexe, *cexe;
  
      tcl_lock = PyThread_allocate_lock();
      if (tcl_lock == NULL)
***************
*** 3409,3426 ****
      m = PyModule_Create(&_tkintermodule);
      if (m == NULL)
          return NULL;
  
!     o = PyErr_NewException("_tkinter.TclError", NULL, NULL);
!     if (o == NULL) {
          Py_DECREF(m);
          return NULL;
      }
-     if (PyModule_AddObject(m, "TclError", Py_NewRef(o))) {
-         Py_DECREF(o);
-         Py_DECREF(m);
-         return NULL;
-     }
-     Tkinter_TclError = o;
  
      if (PyModule_AddIntConstant(m, "READABLE", TCL_READABLE)) {
          Py_DECREF(m);
--- 3439,3453 ----
      m = PyModule_Create(&_tkintermodule);
      if (m == NULL)
          return NULL;
+ #ifdef Py_GIL_DISABLED
+     PyUnstable_Module_SetGIL(m, Py_MOD_GIL_NOT_USED);
+ #endif
  
!     Tkinter_TclError = PyErr_NewException("_tkinter.TclError", NULL, NULL);
!     if (PyModule_AddObjectRef(m, "TclError", Tkinter_TclError)) {
          Py_DECREF(m);
          return NULL;
      }
  
      if (PyModule_AddIntConstant(m, "READABLE", TCL_READABLE)) {
          Py_DECREF(m);
***************
*** 3467,3513 ****
          return NULL;
      }
  
!     o = PyType_FromSpec(&Tkapp_Type_spec);
!     if (o == NULL) {
          Py_DECREF(m);
          return NULL;
      }
-     if (PyModule_AddObject(m, "TkappType", o)) {
-         Py_DECREF(o);
-         Py_DECREF(m);
-         return NULL;
-     }
-     Tkapp_Type = o;
  
!     o = PyType_FromSpec(&Tktt_Type_spec);
!     if (o == NULL) {
          Py_DECREF(m);
          return NULL;
      }
-     if (PyModule_AddObject(m, "TkttType", o)) {
-         Py_DECREF(o);
-         Py_DECREF(m);
-         return NULL;
-     }
-     Tktt_Type = o;
  
!     o = PyType_FromSpec(&PyTclObject_Type_spec);
!     if (o == NULL) {
          Py_DECREF(m);
          return NULL;
      }
-     if (PyModule_AddObject(m, "Tcl_Obj", o)) {
-         Py_DECREF(o);
-         Py_DECREF(m);
-         return NULL;
-     }
-     PyTclObject_Type = o;
  
  
      /* This helps the dynamic loader; in Unicode aware Tcl versions
         it also helps Tcl find its encodings. */
!     uexe = PyUnicode_FromWideChar(Py_GetProgramName(), -1);
!     if (uexe) {
          cexe = PyUnicode_EncodeFSDefault(uexe);
          if (cexe) {
  #ifdef MS_WINDOWS
--- 3494,3522 ----
          return NULL;
      }
  
!     Tkapp_Type = PyType_FromSpec(&Tkapp_Type_spec);
!     if (PyModule_AddObjectRef(m, "TkappType", Tkapp_Type)) {
          Py_DECREF(m);
          return NULL;
      }
  
!     Tktt_Type = PyType_FromSpec(&Tktt_Type_spec);
!     if (PyModule_AddObjectRef(m, "TkttType", Tktt_Type)) {
          Py_DECREF(m);
          return NULL;
      }
  
!     PyTclObject_Type = PyType_FromSpec(&PyTclObject_Type_spec);
!     if (PyModule_AddObjectRef(m, "Tcl_Obj", PyTclObject_Type)) {
          Py_DECREF(m);
          return NULL;
      }
  
  
      /* This helps the dynamic loader; in Unicode aware Tcl versions
         it also helps Tcl find its encodings. */
!     uexe = PySys_GetObject("executable");  // borrowed reference
!     if (uexe && PyUnicode_Check(uexe)) {   // sys.executable can be None
          cexe = PyUnicode_EncodeFSDefault(uexe);
          if (cexe) {
  #ifdef MS_WINDOWS
***************
*** 3546,3552 ****
  #endif /* MS_WINDOWS */
          }
          Py_XDECREF(cexe);
-         Py_DECREF(uexe);
      }
  
      if (PyErr_Occurred()) {
--- 3555,3560 ----
