diff -r -c orig/Mac/Makefile new/Mac/Makefile
*** orig/Mac/Makefile	2020-12-31 10:17:07.000000000 -0600
--- new/Mac/Makefile	2020-12-31 08:53:01.000000000 -0600
***************
*** 8,20 ****
  ENSUREPIP=upgrade
  builddir = ..
  srcdir=.
! prefix=/Library/Frameworks/Python.framework/Versions/3.9
  exec_prefix=${prefix}
  LIBDEST=$(prefix)/lib/python$(VERSION)
  RUNSHARED=DYLD_FRAMEWORK_PATH=/Users/culler/MacPacs/Python/python-3.9.0
  BUILDEXE=.exe
  BUILDPYTHON=$(builddir)/python$(BUILDEXE)
! DESTDIR=
  LDFLAGS=
  FRAMEWORKUNIXTOOLSPREFIX=/usr/local
  PYTHONFRAMEWORK=Python
--- 8,20 ----
  ENSUREPIP=upgrade
  builddir = ..
  srcdir=.
! prefix=$(DESTDIR)/Python.framework/Versions/3.9
  exec_prefix=${prefix}
  LIBDEST=$(prefix)/lib/python$(VERSION)
  RUNSHARED=DYLD_FRAMEWORK_PATH=/Users/culler/MacPacs/Python/python-3.9.0
  BUILDEXE=.exe
  BUILDPYTHON=$(builddir)/python$(BUILDEXE)
! #DESTDIR=
  LDFLAGS=
  FRAMEWORKUNIXTOOLSPREFIX=/usr/local
  PYTHONFRAMEWORK=Python
diff -r -c orig/Makefile new/Makefile
*** orig/Makefile	2020-12-31 10:16:53.000000000 -0600
--- new/Makefile	2020-12-31 08:52:48.000000000 -0600
***************
*** 19,24 ****
--- 19,26 ----
  #
  # See also the section "Build instructions" in the README file.
  
+ DESTDIR=/Users/culler/MacPacs/Python/dist
+ 
  # === Variables set by makesetup ===
  
  MODBUILT_NAMES=      posix  errno  pwd  _sre  _codecs  _weakref  _functools  _operator  _collections  _abc  itertools  atexit  _signal  _stat  time  _thread  _locale  _io  faulthandler  _tracemalloc  _peg_parser  _symtable  xxsubtype
***************
*** 198,206 ****
  ENSUREPIP=      upgrade
  
  # OpenSSL options for setup.py so sysconfig can pick up AC_SUBST() vars.
! OPENSSL_INCLUDES=
! OPENSSL_LIBS=
! OPENSSL_LDFLAGS=
  
  # Default zoneinfo.TZPATH. Added here to expose it in sysconfig.get_config_var
  TZPATH=/usr/share/zoneinfo:/usr/lib/zoneinfo:/usr/share/lib/zoneinfo:/etc/zoneinfo
--- 200,208 ----
  ENSUREPIP=      upgrade
  
  # OpenSSL options for setup.py so sysconfig can pick up AC_SUBST() vars.
! OPENSSL_INCLUDES=-I/Library/Frameworks/OpenSSL.framework/Versions/1.1/Resources/include
! OPENSSL_LIBS=-lssl -lcrypto
! OPENSSL_LDFLAGS=-L/Library/Frameworks/OpenSSL.framework/Versions/1.1/Resources/lib
  
  # Default zoneinfo.TZPATH. Added here to expose it in sysconfig.get_config_var
  TZPATH=/usr/share/zoneinfo:/usr/lib/zoneinfo:/usr/share/lib/zoneinfo:/etc/zoneinfo
***************
*** 1744,1750 ****
  
  # This installs the IDE, the Launcher and other apps into /Applications
  frameworkinstallapps:
! 	cd Mac && $(MAKE) installapps DESTDIR="$(DESTDIR)"
  
  # Build the bootstrap executable that will spawn the interpreter inside
  # an app bundle within the framework.  This allows the interpreter to
--- 1746,1752 ----
  
  # This installs the IDE, the Launcher and other apps into /Applications
  frameworkinstallapps:
! 	echo Not running: cd Mac && $(MAKE) installapps DESTDIR="$(DESTDIR)"
  
  # Build the bootstrap executable that will spawn the interpreter inside
  # an app bundle within the framework.  This allows the interpreter to
